README.rst
Chroma
======

Chroma is a Python module for handling colors with ease.

Manipulating colors can quickly escalate into a tedious and complicated task, particularly when you become concerned with color systems beyond RGB. Chroma is here to provide a simple API to do the heavy lifting, so that you can stay focused on the important parts of your projects.

Before you ask, Chroma is BSD licensed, available on `Github <https://github.com/seenaburns/Chroma>`_ and PyPI.

Features
--------

- Hex (#rrggbb, #rrggbbaa)
- RGB
- HLS
- HSV
- CMY, CMYK
- Alpha
- Color Blending: additive and subtractive mixing

Roadmap
-------

- Coordinates (sRGB, YIQ, CIE, LAB and more)
- Color harmonies: complementary, analogous, triad
- Color difference
- Color detection in images

Quickstart
==========

Getting started with the power of Chroma is meant to be straightforward
::

    import chroma

    # Create a color object
    color = chroma.Color('#00FF00')

    # Handling different color systems
    color.cmy = (0.3, 0.7, 0.8)
    color.rgb    # (0.7, 0.3, 0.2)
    color.hls    # (0.0333, 0.45, 0.5556)

    # Alpha
    color.alpha = 0.5
    color.hsv    # (0.03333, 0.7143, 0.7, 0.5)

    # Color blending
    color + chroma.Color("#FF00FF")
    # #FF4DFF

And there you have it. The `documentation <https://chroma.readthedocs.org/en/latest/>`_ describes Chroma's functionality and usage in detail.

Installation
------------

Installation is as easy as
::

    pip install chroma

Or if you're an easy_install-er
::

    easy_install chroma

Chroma does not yet support Python 3, but, if you're interested, contribute!

Contribute
==========

Chroma is under active development and could use your support. Even bug reports, feature suggestions and feedback can help push Chroma forward in the right direction.

Chroma is hosted on `Github <https://github.com/seenaburns/Chroma>`_ and there are a number of ideas of where to start in the `issues section <https://github.com/seenaburns/Chroma/issues>`_.


HISTORY.rst
History
=======

0.2.0 (2013-02-01)
------------------

(Breaks backwards compatibility)

- CMY, CMYK Support
- Color Blending (Additive and Subtractive Mixing)
- Comparison (eq, ne)
- Remove color coordinate properties (direct modification)

Major Bug Fixes

- Hex rounding
- Force alpha between 0 and 1

Other

- Hue (HLS, HSV) in degrees not percent of degrees
- RGB256 alpha in range 0-1, not 0-255
- Force Hex to output in uppercase
- ValueError on bad input
- Extensive documentation on Read the Docs

0.1.3 (2013-01-01)
------------------
- Direct modification of color coordinates

0.1.2 (2012-12-28)
------------------
- Remove HLS256, HSV256 (critical bug)
- Alpha support
- Add RGB, RGB256, HLS, HSV as formats for Color initialization
- Bug fixes

0.1.1 (2012-12-16)
------------------
- HLS, HSV Support
- API changes (setter methods)
- Bug Fixes
- Update to README

0.1.0 (2012-12-15)
------------------
- Initial Release
- RGB and Hex support


tests/test_chroma.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Chroma Test Suite
~~~~~~~~~~~~~~~~~~~

A series of unittests to check the basic functionality of the Chroma API.

"""

import unittest
import argparse

# Path hack. (for importing)
import sys
import os
sys.path.insert(0, os.path.abspath('..'))
sys.path.insert(0, os.path.abspath('.'))
import chroma

class ChromaTestSuite(unittest.TestCase):
    def setUp(self):
        self.c1 = chroma.Color('#335577')
        self.c2 = chroma.Color('#446688')
        self.c3 = chroma.Color('#555555')

    def assertTupleAlmostEqual(self, t1, t2):
        for x, y in zip(t1, t2):
            self.assertEqual(round(x - y, 2), 0)

    def test_color_initialization(self):
        """Test construction of color object"""

        white = chroma.Color('#FFFFFF').hex

        self.assertEqual(chroma.Color((1,1,1), 'RGB').hex, white)
        self.assertEqual(chroma.Color((255, 255, 255), 'RGB256').hex, white)
        self.assertEqual(chroma.Color((0, 1, 1), 'HLS').hex, white)
        self.assertEqual(chroma.Color((0, 0, 1), 'HSV').hex, white)
        self.assertEqual(chroma.Color((0, 0, 0, 0), 'CMY').hex, white)
        self.assertEqual(chroma.Color((0, 0, 0, 0), 'CMYK').hex, white)

        # Test ValueError too
        self.assertRaises(ValueError, chroma.Color, (0, 0, 0), 'ERROR')

    def test_comparison_methods(self):
        """Test equality and inequality"""

        # Equality
        self.assertFalse(self.c1 == self.c2)
        self.assertTrue(self.c1 == chroma.Color('#335577'))

        # Inequality
        self.assertTrue(self.c1 != self.c2)
        self.assertFalse(self.c1 != chroma.Color('#335577'))

    def test_system_conversion(self):
        """Test conversion between systems"""

        # Use Wolfram Alpha as external reference
        # Everything -> RGB
        self.assertTupleAlmostEqual(chroma.Color((0.2, 0.3333, 0.4667), 'RGB').rgb, self.c1.rgb)
        self.assertTupleAlmostEqual(chroma.Color((51, 85, 119), 'RGB256').rgb, self.c1.rgb)
        self.assertTupleAlmostEqual(chroma.Color((210, 0.3333, 0.40), 'HLS').rgb, self.c1.rgb)
        self.assertTupleAlmostEqual(chroma.Color((210, 0.57, 0.466), 'HSV').rgb, self.c1.rgb)
        self.assertTupleAlmostEqual(chroma.Color((0.8, 0.67, 0.53), 'CMY').rgb, self.c1.rgb)
        self.assertTupleAlmostEqual(chroma.Color((0.57, 0.29, 0, 0.53), 'CMYK').rgb, self.c1.rgb)

        # RGB -> Everything
        self.assertTupleAlmostEqual(chroma.Color((0.2, 0.3333, 0.4667), 'RGB').rgb, self.c1.rgb)
        self.assertTupleAlmostEqual(chroma.Color((51, 85, 119), 'RGB256').rgb256, self.c1.rgb256)
        self.assertTupleAlmostEqual(chroma.Color((210, 0.3333, 0.40), 'HLS').hls, self.c1.hls)
        self.assertTupleAlmostEqual(chroma.Color((210, 0.57, 0.466), 'HSV').hsv, self.c1.hsv)
        self.assertTupleAlmostEqual(chroma.Color((0.8, 0.67, 0.53), 'CMY').cmy, self.c1.cmy)
        self.assertTupleAlmostEqual(chroma.Color((0.57, 0.29, 0, 0.53), 'CMYK').cmyk, self.c1.cmyk)

    def test_bad_input(self):
        """Test input that goes beyond color system bounds"""
        # Upper bound
        self.assertEqual(chroma.Color((10, -3, 0.5), 'RGB').rgb, chroma.Color((1, 0, 0.5), 'RGB').rgb)
        self.assertEqual(chroma.Color((300, -3, 50), 'RGB256').rgb256, chroma.Color((255, 0, 50), 'RGB256').rgb256)
        self.assertEqual(chroma.Color((400, -3, 10), 'HLS').hls, chroma.Color((360, 0, 1), 'HLS').hls)
        self.assertEqual(chroma.Color((-10, 40, -1), 'HLS').hls, chroma.Color((0, 1, 0), 'HLS').hls)
        self.assertEqual(chroma.Color((400, -3, 10), 'HSV').hsv, chroma.Color((360, 0, 1), 'HSV').hsv)
        self.assertEqual(chroma.Color((-10, 40, -1), 'HSV').hsv, chroma.Color((0, 1, 0), 'HSV').hsv)
        self.assertEqual(chroma.Color((10, -3, 0.5), 'CMY').cmy, chroma.Color((1, 0, 0.5), 'CMY').cmy)
        self.assertEqual(chroma.Color((10, -3, 0.5, 3), 'CMYK').cmyk, chroma.Color((1, 0, 0.5, 1), 'CMYK').cmyk)

        # Test alpha
        self.c1.rgb = (10, -1, 0.5, 10)
        self.assertEqual(self.c1.rgb, (1, 0, 0.5, 1))
        self.c1.rgb = (10, -1, 0.5, -10)
        self.assertEqual(self.c1.rgb, (1, 0, 0.5, 0))

    def test_alpha(self):
        """Test alpha support / no-support with various color systems"""
        alpha_value = 0.5
        hex_value = self.c1.hex + '80'

        # Test off
        self.assertEqual(self.c1.alpha, None)

        self.c1.alpha = alpha_value

        # Test color systems
        # Test if it is there, and the value is still correct
        self.assertEqual(self.c1.alpha, alpha_value)
        self.assertEqual(self.c1.hex, hex_value)
        self.assertEqual(self.c1.rgb[3], alpha_value)
        self.assertEqual(self.c1.rgb256[3], alpha_value)
        self.assertEqual(self.c1.hls[3], alpha_value)
        self.assertEqual(self.c1.hsv[3], alpha_value)

        # Test no alpha
        self.assertEqual(len(self.c1.cmy), 3)
        self.assertEqual(len(self.c1.cmyk), 4)

        # Test off after
        self.c1.alpha = None
        self.assertEqual(self.c1.alpha, None)
        self.assertEqual(len(self.c1.hex), 7)
        self.assertEqual(len(self.c1.rgb), 3)
        self.assertEqual(len(self.c1.rgb256), 3)
        self.assertEqual(len(self.c1.hls), 3)
        self.assertEqual(len(self.c1.hsv), 3)

    def test_color_blending(self):
        """Test additive and subtractive mixing"""

        # Additive mixing
        self.assertEqual(chroma.Color("#77BBFF"), self.c1 + self.c2)
        self.assertEqual(chroma.Color("#99BBDD"), self.c3 + self.c2)
        self.assertEqual(chroma.Color("#FFFFFF"), chroma.Color("#FFFFFF") + chroma.Color("#000000"))
        self.assertEqual(self.c1, self.c1 + chroma.Color("#000000"))

        # Subtractive mixing
        self.assertEqual(chroma.Color("#FFFF00"), chroma.Color("#FFFFFF") - chroma.Color("#FFFF00"))
        self.assertEqual(chroma.Color("#00FF00"), chroma.Color("#FFFF00") - chroma.Color("#00FFFF"))


if __name__ == '__main__':
    unittest.main()


chroma/core.py
# -*- coding: utf-8 -*-

"""
chroma.core
~~~~~~~~~~~~~

Provides Color object

"""

# Tuples are assumed to be passed and returned

import colorsys

class Color(object):
    """
    Chroma Color object stores 'color value' to be given in any format
    by one of the properties
    """
    def __init__(self, color_value = '#FFFFFF', format = 'HEX'):
        # self.color is main storage for color format (tuple in RGB float form)
        # HEX input takes string, RGB / HLS / HSV take tuples

        self.color = (1.0, 1.0, 1.0)
        # If alpha is None, it is unset and assumed to be RGB
        # If non-negative, it has been set and use RGBA, HLSA, etc
        # _alpha used internally
        self._alpha = None

        if format.upper() == 'HEX':
            self.rgb = self._rgb_from_hex(color_value)
        elif format.upper() == 'RGB':
            self.rgb = color_value
        elif format.upper() == 'RGB256':
            self.rgb256 = color_value
        elif format.upper() == 'HLS':
            self.hls = color_value
        elif format.upper() == 'HSV':
            self.hsv = color_value
        elif format.upper() == 'CMY':
            self.cmy = color_value
        elif format.upper() == 'CMYK':
            self.cmyk = color_value
        else:
            raise ValueError('Unsupported chroma.Color format: %s' % (format))

    # Color equality: difference is less than a tolerance
    # Use hex as the test for equals, as it is the greatest resolution without rounding issues
    def __eq__(self, other):
        return self.hex == other.hex

    def __ne__(self, other):
        return not (self == other)

    # Additive / subtractive mixing
    def __add__(self, other):
        return self.additive_mix(other)

    def __radd__(self, other):
        return other.additive_mix(self)

    def __sub__(self, other):
        return self.subtractive_mix(other)

    def __rsub__(self, other):
        return other.subtractive_mix(self)

    # Representation
    def __str__(self):
        return self.hex

    def __repr__(self):
        return self.__str__()

    #
    # Properties
    #

    # RGB
    # RGB is used as base, other formats will modify input into RGB and invoke
    # RGB getters / setters
    @property
    def rgb(self):
        return self._append_alpha_if_necessary(self.color)

    @property
    def rgb256(self):
        rgb256 = tuple(map(lambda x: int(round(x*255)), self.color))
        return self._append_alpha_if_necessary(rgb256)

    @rgb.setter
    def rgb(self, color_tuple):
        """Used as main setter (rgb256, hls, hls256, hsv, hsv256)"""
        # Check bounds
        self.color = tuple(map(self._apply_float_bounds, color_tuple[:3]))

        # Include alpha if necessary
        if len(color_tuple) > 3:
            self.alpha = self._apply_float_bounds(color_tuple[3])

    @rgb256.setter
    def rgb256(self, color_tuple):
        self.rgb = map(lambda x: x / 255.0, color_tuple)

    # HLS
    @property
    def hls(self):
        """
        HLS: (Hue°, Lightness%, Saturation%)
        Hue given as percent of 360, Lightness and Saturation given as percent
        """
        r, g, b = self.color
        hls = colorsys.rgb_to_hls(r, g, b)
        hls = (int(round(hls[0] * 360)), hls[1], hls[2])
        return self._append_alpha_if_necessary(hls)

    @hls.setter
    def hls(self, color_tuple):
        h, l, s = (self._apply_float_bounds(color_tuple[0]/360.0),
                   self._apply_float_bounds(color_tuple[1]),
                   self._apply_float_bounds(color_tuple[2]))
        rgb = colorsys.hls_to_rgb(h, l, s)

        # Append alpha if included
        if len(color_tuple) > 3:
            rgb += (color_tuple[3],)

        self.rgb = rgb

    # HSV
    @property
    def hsv(self):
        """
        HSV: (Hue°, Saturation%, Value%)
        Hue given as percent of 360, Saturation and Value given as percent
        """
        r, g, b = self.color
        hsv = colorsys.rgb_to_hsv(r, g, b)
        hsv = (int(round(hsv[0] * 360)), hsv[1], hsv[2])
        return self._append_alpha_if_necessary(hsv)

    @hsv.setter
    def hsv(self, color_tuple):
        h, s, v = (self._apply_float_bounds(color_tuple[0]/360.0),
                   self._apply_float_bounds(color_tuple[1]),
                   self._apply_float_bounds(color_tuple[2]))
        rgb = colorsys.hsv_to_rgb(h, s, v)

        # Append alpha if included
        if len(color_tuple) > 3:
            rgb += (color_tuple[3],)

        self.rgb = rgb

    # CMY / CMYK
    @property
    def cmy(self):
        """
        CMY: returned in range 0.0 - 1.0
        CMY is subtractive, e.g. black: (1, 1, 1), white (0, 0, 0)
        """
        r, g, b = self.color
        c = 1 - r
        m = 1 - g
        y = 1 - b

        return (c, m, y)

    @cmy.setter
    def cmy(self, color_tuple):
        c, m, y = tuple(map(lambda x: self._apply_float_bounds(x), color_tuple))[:3]
        r = 1 - c
        g = 1 - m
        b = 1 - y
        self.rgb = (r, g, b)

    @property
    def cmyk(self):
        """CMYK: all returned in range 0.0 - 1.0"""
        c, m, y = self.cmy
        k = min(c, m, y)

        # Handle division by zero in case of black = 1
        if k != 1:
            c = (c - k) / (1 - k)
            m = (m - k) / (1 - k)
            y = (y - k) / (1 - k)
        else:
            c, m, y = 1, 1, 1

        cmyk = (c, m, y, k)

        # Apply bound and return
        return tuple(map(lambda x: self._apply_float_bounds(x), cmyk))

    @cmyk.setter
    def cmyk(self, color_tuple):
        c, m, y, k = tuple(map(lambda x: self._apply_float_bounds(x), color_tuple))[:4]
        c = c * (1 - k) + k
        m = m * (1 - k) + k
        y = y * (1 - k) + k
        self.cmy = (c, m, y)

    # HEX
    @property
    def hex(self):
        r = self._float_to_hex(self.color[0])
        g = self._float_to_hex(self.color[1])
        b = self._float_to_hex(self.color[2])
        rgb = '#' + r + g + b

        # Append alpha hex if necessary
        if self.alpha is not None:
            rgb += self._float_to_hex(self.alpha)

        return rgb

    @hex.setter
    def hex(self, color_value):
        self.rgb = self._rgb_from_hex(color_value)

    # Alpha
    @property
    def alpha(self):
        return self._alpha

    @alpha.setter
    def alpha(self, value):
        self._alpha = self._apply_float_bounds(value)

    #
    # Color Functions
    # Note: color functions return a Color object when called directly
    #

    # Additive (Light) Mixing
    def additive_mix(self, other):
        rgb_mix = tuple([rgb1 + rgb2 for rgb1, rgb2 in zip(self.rgb, other.rgb)])
        return Color(rgb_mix, 'RGB')

    # Subtractive (Dye, Multiplicative) Mixing
    def subtractive_mix(self, other):
        cmy_mix = tuple([cmy1 + cmy2 for cmy1, cmy2 in zip(self.cmy, other.cmy)])
        return Color(cmy_mix, 'CMY')

    #
    # INTERNAL
    #
    def _rgb_from_hex(self, color_value):
        hex_value = str(color_value)

        # Remove hash if exists
        if hex_value[0] == '#':
            hex_value = hex_value[1:]

        # Check length
        # 6: 6 digit hex
        # 8: 6 digit hex + alpha
        if len(hex_value) not in [6, 8]:
            raise ValueError('Invalid Hex Input: %s' % (color_value))

        # Return rgb from hex
        try:
            rgb =  (int(hex_value[0:2], 16) / 255.0,
                    int(hex_value[2:4], 16) / 255.0,
                    int(hex_value[4:6], 16) / 255.0)

            # Append alpha if exists
            if len(hex_value) == 8:
                rgb += (int(hex_value[6:8], 16) / 255.0,)

            return rgb
        except Exception, e:
            raise ValueError('Invalid Hex Input: %s' % (color_value))

    def _float_to_hex(self, float_value):
        # Convert from float to in to hex number, remove '0x'
        int_value = int(round(float_value*255))
        hex_value = hex(int_value)[2:]

        # If hex is only one digit, pad with 0
        if len(hex_value) == 1:
            hex_value = '0' + hex_value

        return hex_value.upper()

    def _apply_float_bounds(self, coordinate):
        """Assure coordinate is a float between 0 to 1"""
        # Skip None for Alpha
        if coordinate == None:
            return None

        if coordinate < 0.0:
            return 0.0
        elif coordinate > 1.0:
            return 1.0

        return float(coordinate)

    def _append_alpha_if_necessary(self, color_tuple):
        """Return color_tuple with alpha if self.alpha is not None"""
        if self.alpha is not None:
            return color_tuple + (self.alpha,)
        return color_tuple

